src/Interpreters/InterpreterCreateQuery.cpp

```diff
@@ -723,8 +723,10 @@ InterpreterCreateQuery::TableProperties InterpreterCreateQuery::getTableProperti
     if (create.storage && create.storage->engine)
     {
         auto source_access_type = StorageFactory::instance().getSourceAccessType(create.storage->engine->name);
-        if (source_access_type != AccessType::NONE)
+        const auto & access_control = getContext()->getAccessControl();
+        if (source_access_type != AccessType::NONE && !access_control.doesTableEnginesRequireGrant())
             getContext()->checkAccess(source_access_type);
+        getContext()->checkAccess(AccessType::TABLE_ENGINE, create.storage->engine->name);
```

Comment:
We can move this quite access-specific logic to `ContextAccess` and leave only the following here:
```
    if (create.storage && create.storage->engine)
        getContext()->checkAccess(AccessType::TABLE_ENGINE, create.storage->engine->name);
```

Because `InterpreterCreateQuery` is quite a complex class already and it's better not to burden it too much with access-specific internal logic. Function `ContextAccess::addImplicitAccessRights()` can add `TABLE_ENGINE` grant if `access_control.doesTableEnginesRequireGrant()` is false and if the table engine is one of the source types and this source type is granted.

Reply:
The original code here is to check source types, where I modified to avoid checking source types again if table engine grant is enabled, as checking table engine grant is enough. So here I assume if table engine is granted, then source access type is granted as well according to our last discussion [here](https://github.com/ClickHouse/ClickHouse/pull/56945#issuecomment-1908642898). I am not assuming source access type is granted, then table engine is granted. Is this requirement still valid?

Reply:
@vitlibar Can I assume table engine is granted then source access type is granted, but not vice versa?

Reply:
@vitlibar I somehow misunderstood last time. Now treated source and table engine equally in grant, and added code in `addImplicitAccessRights`. But I think we still need to differentiate the cases whether grant of table engine is enabled here. Otherwise, in the case of not enabled, only check table engine is not enough. (though we can change the interface of some upper level check funcs, but not worthwhile just for this specific case). WDYT?